### PR TITLE
Added a collection lazy() method

### PR DESCRIPTION
**Pull request by lorenzo**
Since cake 3.6 collections are not entirely lazy anymore. This is
because in the greate majority of cases, running the computations
eagerly is faster than doing it on demand.

This assumption fails, though, when the computations are very intensive
or when you actually only need a handful of elements anyways. In order
to get this feature back, I'm introducing the `lazy()` method. It will
make sure that computations only happen on demand.

I feel that this give more power to the developer who's trying to find
the most performant way of traversing a set of results.

### CODE REVIEW

**Comment by lorenzo** [comment]
I'll try to improve the code coverage a bit and then will merge

**Comment by ravage84** [comment]
> Since cake 3.6 collections are not entirely lazy anymore. This is
because in the greate majority of cases, running the computations
eagerly is faster than doing it on demand.

@lorenzo  are there deeper implications because of this? Did we update the docs about this?

File: src/Collection/CollectionInterface.php [hunk 1]
```diff
@@ -783,6 +783,16 @@ public function jsonSerialize();
      */
     public function compile($preserveKeys = true);
 
+    /**
+     * Returns a new collection where any operations chained after it are guaranteed
+     * to be run lazily. That is, elements will be yieleded one at a time.
+     *
+     * A lazy collection can only be iterated once. A second attempt results in an error.
```
Comment (ravage84) [inline src/Collection/CollectionInterface.php]: :+1:  Good note! This should be included in the Cookbook, too.

Comment (ravage84) [reply src/Collection/CollectionInterface.php]: Would there be any value in a test verifiing this behavior? (if we aren't doing that already)

File: tests/TestCase/Collection/CollectionTest.php [hunk 2]
```diff
@@ -2632,4 +2632,21 @@ public function testArrayIteratorExtend()
         $this->assertTrue($newIterator->checkValues());
         $this->assertCount(3, $newIterator->toArray());
     }
+
+    /**
+     * Tests that elements in a lazy collection are only fetched once
```
Comment (ravage84) [inline tests/TestCase/Collection/CollectionTest.php]: I think this title is a bit misleading.  While implicitly true, we deliberately test that the the filter callable will not be called for the lazy collection without exporting to array for example.

Shouldn't we add a test that verifies the actual expected behavior mention in this title?

**Review by ravage84 — CHANGES_REQUESTED** [review]
> I feel that this give more power to the developer who's trying to find
the most performant way of traversing a set of results.

Agree, choice is most often a good thing.

File: tests/TestCase/Collection/CollectionTest.php (hunk 2 above)
Comment (lorenzo) [reply tests/TestCase/Collection/CollectionTest.php]: The test is actually verifying said behaviour. If the collection wasn't lazy, then the filters would e called immediately

Comment (ravage84) [reply tests/TestCase/Collection/CollectionTest.php]: I agree. It's just a wording thing, I guess.
How about rewording to: "Tests that elements in a lazy collection are not fetched immediately"?

Comment (lorenzo) [reply tests/TestCase/Collection/CollectionTest.php]: sounds good to me

**Review by ravage84 — APPROVED** [review]
